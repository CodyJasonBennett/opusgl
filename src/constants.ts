export const DATA_TYPES = {
  UNSIGNED_SHORT: 0x1403,
  FLOAT: 0x1406,
  FLOAT_VEC2: 0x8b50,
  FLOAT_VEC3: 0x8b51,
  FLOAT_VEC4: 0x8b52,
  BOOL: 0x8b56,
  INT: 0x1404,
  SAMPLER_2D: 0x8b5e,
  SAMPLER_CUBE: 0x8b60,
  BOOL_VEC2: 0x8b57,
  INT_VEC2: 0x8b53,
  BOOL_VEC3: 0x8b58,
  INT_VEC3: 0x8b54,
  BOOL_VEC4: 0x8b59,
  INT_VEC4: 0x8b55,
  FLOAT_MAT2: 0x8b5a,
  FLOAT_MAT3: 0x8b5b,
  FLOAT_MAT4: 0x8b5c,
} as const

export const GL = {
  CLEAR_COLOR: 0x00004000,
  CLEAR_DEPTH: 0x00000100,
  BUFFER_TYPE: 0x8892,
  BUFFER_INDEX_TYPE: 0x8893,
  BUFFER_USAGE: 0x88e4,
  ATTRIBUTE_TYPE: DATA_TYPES.FLOAT,
  ATTRIBUTE_INDEX_TYPE: DATA_TYPES.UNSIGNED_SHORT,
  EXTENSIONS_DEPTH: 0x0b71,
  EXTENSIONS_SCISSOR: 0x0c11,
  PROGRAM_COMPILE_STATUS: 0x8b81,
  PROGRAM_LINK_STATUS: 0x8b82,
  PROGRAM_UNIFORMS: 0x8b86,
  SHADER_VERTEX: 0x8b31,
  SHADER_FRAGMENT: 0x8b30,
} as const

export const DRAW_MODES = {
  POINTS: 0x0000,
  LINES: 0x0001,
  LINE_STRIP: 0x0003,
  TRIANGLES: 0x0004,
  TRIANGLE_STRIP: 0x0005,
} as const

export const GPU_DRAW_MODES = {
  POINTS: 'point-list',
  LINES: 'line-list',
  LINE_STRIP: 'line-strip',
  TRIANGLES: 'triangle-list',
  TRIANGLE_STRIP: 'triangle-strip',
} as const

export const SHADER_TEMPLATES = {
  vertex: '#version 300 es\nprecision highp float;\nprecision highp int;',
  fragment: '#version 300 es\nprecision highp float;\nprecision highp int;',
} as const

export const COLORS = {
  aliceblue: 0xf0f8ff,
  antiquewhite: 0xfaebd7,
  aqua: 0x00ffff,
  aquamarine: 0x7fffd4,
  azure: 0xf0ffff,
  beige: 0xf5f5dc,
  bisque: 0xffe4c4,
  black: 0x000000,
  blanchedalmond: 0xffebcd,
  blue: 0x0000ff,
  blueviolet: 0x8a2be2,
  brown: 0xa52a2a,
  burlywood: 0xdeb887,
  cadetblue: 0x5f9ea0,
  chartreuse: 0x7fff00,
  chocolate: 0xd2691e,
  coral: 0xff7f50,
  cornflowerblue: 0x6495ed,
  cornsilk: 0xfff8dc,
  crimson: 0xdc143c,
  cyan: 0x00ffff,
  darkblue: 0x00008b,
  darkcyan: 0x008b8b,
  darkgoldenrod: 0xb8860b,
  darkgray: 0xa9a9a9,
  darkgreen: 0x006400,
  darkgrey: 0xa9a9a9,
  darkkhaki: 0xbdb76b,
  darkmagenta: 0x8b008b,
  darkolivegreen: 0x556b2f,
  darkorange: 0xff8c00,
  darkorchid: 0x9932cc,
  darkred: 0x8b0000,
  darksalmon: 0xe9967a,
  darkseagreen: 0x8fbc8f,
  darkslateblue: 0x483d8b,
  darkslategray: 0x2f4f4f,
  darkslategrey: 0x2f4f4f,
  darkturquoise: 0x00ced1,
  darkviolet: 0x9400d3,
  deeppink: 0xff1493,
  deepskyblue: 0x00bfff,
  dimgray: 0x696969,
  dimgrey: 0x696969,
  dodgerblue: 0x1e90ff,
  firebrick: 0xb22222,
  floralwhite: 0xfffaf0,
  forestgreen: 0x228b22,
  fuchsia: 0xff00ff,
  gainsboro: 0xdcdcdc,
  ghostwhite: 0xf8f8ff,
  gold: 0xffd700,
  goldenrod: 0xdaa520,
  gray: 0x808080,
  green: 0x008000,
  greenyellow: 0xadff2f,
  grey: 0x808080,
  honeydew: 0xf0fff0,
  hotpink: 0xff69b4,
  indianred: 0xcd5c5c,
  indigo: 0x4b0082,
  ivory: 0xfffff0,
  khaki: 0xf0e68c,
  lavender: 0xe6e6fa,
  lavenderblush: 0xfff0f5,
  lawngreen: 0x7cfc00,
  lemonchiffon: 0xfffacd,
  lightblue: 0xadd8e6,
  lightcoral: 0xf08080,
  lightcyan: 0xe0ffff,
  lightgoldenrodyellow: 0xfafad2,
  lightgray: 0xd3d3d3,
  lightgreen: 0x90ee90,
  lightgrey: 0xd3d3d3,
  lightpink: 0xffb6c1,
  lightsalmon: 0xffa07a,
  lightseagreen: 0x20b2aa,
  lightskyblue: 0x87cefa,
  lightslategray: 0x778899,
  lightslategrey: 0x778899,
  lightsteelblue: 0xb0c4de,
  lightyellow: 0xffffe0,
  lime: 0x00ff00,
  limegreen: 0x32cd32,
  linen: 0xfaf0e6,
  magenta: 0xff00ff,
  maroon: 0x800000,
  mediumaquamarine: 0x66cdaa,
  mediumblue: 0x0000cd,
  mediumorchid: 0xba55d3,
  mediumpurple: 0x9370db,
  mediumseagreen: 0x3cb371,
  mediumslateblue: 0x7b68ee,
  mediumspringgreen: 0x00fa9a,
  mediumturquoise: 0x48d1cc,
  mediumvioletred: 0xc71585,
  midnightblue: 0x191970,
  mintcream: 0xf5fffa,
  mistyrose: 0xffe4e1,
  moccasin: 0xffe4b5,
  navajowhite: 0xffdead,
  navy: 0x000080,
  oldlace: 0xfdf5e6,
  olive: 0x808000,
  olivedrab: 0x6b8e23,
  orange: 0xffa500,
  orangered: 0xff4500,
  orchid: 0xda70d6,
  palegoldenrod: 0xeee8aa,
  palegreen: 0x98fb98,
  paleturquoise: 0xafeeee,
  palevioletred: 0xdb7093,
  papayawhip: 0xffefd5,
  peachpuff: 0xffdab9,
  peru: 0xcd853f,
  pink: 0xffc0cb,
  plum: 0xdda0dd,
  powderblue: 0xb0e0e6,
  purple: 0x800080,
  rebeccapurple: 0x663399,
  red: 0xff0000,
  rosybrown: 0xbc8f8f,
  royalblue: 0x4169e1,
  saddlebrown: 0x8b4513,
  salmon: 0xfa8072,
  sandybrown: 0xf4a460,
  seagreen: 0x2e8b57,
  seashell: 0xfff5ee,
  sienna: 0xa0522d,
  silver: 0xc0c0c0,
  skyblue: 0x87ceeb,
  slateblue: 0x6a5acd,
  slategray: 0x708090,
  slategrey: 0x708090,
  snow: 0xfffafa,
  springgreen: 0x00ff7f,
  steelblue: 0x4682b4,
  tan: 0xd2b48c,
  teal: 0x008080,
  thistle: 0xd8bfd8,
  tomato: 0xff6347,
  turquoise: 0x40e0d0,
  violet: 0xee82ee,
  wheat: 0xf5deb3,
  white: 0xffffff,
  whitesmoke: 0xf5f5f5,
  yellow: 0xffff00,
  yellowgreen: 0x9acd32,
} as const
