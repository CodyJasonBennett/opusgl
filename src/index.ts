export * from './core/Camera'
export * from './core/Geometry'
export * from './core/Material'
export * from './core/Mesh'
export * from './core/Object3D'
export * from './core/Program'
export * from './core/Renderer'
export * from './core/Scene'
export * from './core/Texture'

export * from './math/Color'
export * from './math/Euler'
export * from './math/Matrix3'
export * from './math/Matrix4'
export * from './math/Quaternion'
export * from './math/Vector2'
export * from './math/Vector3'
export * from './math/Vector4'

export * from './cameras/PerspectiveCamera'
export * from './cameras/OrthographicCamera'

export * from './renderers/WebGLRenderer'
export * from './renderers/WebGPURenderer'

export * from './constants'
export * from './utils'
