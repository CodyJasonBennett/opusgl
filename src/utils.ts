import type { Uniform } from './core/Material'

/**
 * Adds line numbers to a string with an optional starting offset.
 */
export const lineNumbers = (source: string, offset = 0): string => source.replace(/^/gm, () => `${offset++}:`)

/**
 * Generates a v4 UUID. Useful for tracking unique objects.
 */
export const uuid = (): string =>
  '00-0-4-1-000'.replace(/\d/g, (s) =>
    (((Math.random() + Number(s)) * 0x10000) >> Number(s)).toString(16).padStart(4, '0').toUpperCase(),
  )

/**
 * Compares two uniforms by keys or reference.
 */
export const uniformsEqual = (a: Uniform, b: Uniform): boolean => {
  // Recursively compare array uniforms
  if (a instanceof Array && b instanceof Array) return a.every((v, i) => uniformsEqual(v, b[i]))
  // Atomically compare literals
  return a === b
}

/**
 * Clones a uniform's value into memory.
 */
export const cloneUniform = (uniform: Uniform, prev?: Uniform): Uniform => {
  // @ts-ignore
  return (prev ? prev.copy?.(uniform) : uniform.clone?.()) ?? uniform
}
