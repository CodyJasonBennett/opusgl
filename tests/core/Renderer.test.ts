import { describe, it, expect } from 'vitest'
import { Renderer, Scene, Mesh, Geometry, Material, Vector2 } from '../../src'

class RendererImpl extends Renderer {
  render() {}
}

describe('core/Renderer', () => {
  it('can sort a list of meshes by visibility', () => {
    const renderer = new RendererImpl()

    const scene = new Scene()
    scene.add(new Mesh(new Geometry(), new Material()))

    const renderList = renderer.sort(scene)

    expect(renderList.length).toBe(1)
  })

  it('can compare uniforms', () => {
    const renderer = new RendererImpl()

    // Compares initial uniforms
    expect(renderer.uniformsEqual(undefined, 0)).toBe(false)
    expect(renderer.uniformsEqual(undefined, undefined)).toBe(true)

    // Compares math classes via #equals
    expect(renderer.uniformsEqual(new Vector2(1), new Vector2(1))).toBe(true)
    expect(renderer.uniformsEqual(new Vector2(1), new Vector2(2))).toBe(false)

    // Atomically compares numbers
    expect(renderer.uniformsEqual(1, 1)).toBe(true)
    expect(renderer.uniformsEqual(1, 2)).toBe(false)
  })

  it('can parse uniform definitions from GLSL', () => {
    const renderer = new RendererImpl()
    const names = renderer.parseUniforms(
      `
      /*
        layout(std140) uniform Uniforms {};
      */

      layout(std140) uniform Uniforms {
        // bool test;
        float time;
        vec3 color;
      };
    `,
      '',
    )

    expect(names).toMatchSnapshot()
  })

  it('can parse uniform definitions from WGSL', () => {
    const renderer = new RendererImpl()
    const names = renderer.parseUniforms(
      `
      /*
        @binding(0) @group(0) var<uniform> uniforms: Foo;
      */

      struct Uniforms {
        // bool test,
        time: f32,
        color: vec3<f32>
      };
      @binding(0) @group(0) var<uniform> uniforms: Uniforms;
    `,
      '',
    )

    expect(names).toMatchSnapshot()
  })
})
