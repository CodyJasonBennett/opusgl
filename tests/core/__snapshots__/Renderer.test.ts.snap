// Vitest Snapshot v1

exports[`core/Renderer > can parse uniform definitions from GLSL 1`] = `
[
  "time",
  "color",
]
`;

exports[`core/Renderer > can parse uniform definitions from WGSL 1`] = `
[
  "time",
  "color",
]
`;
